### PR TITLE
crimson/osd/osd_operations: make the "recover_missing" phase concurrent

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Otherwise, UrgentRecovery requests within the same pg can only be processed sequentially

Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by Matan-B — APPROVED** [review]
This is not strictly a bug, right?

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-06-17_15:38:16-crimson-rados-wip-matanb-crimson-testing-june-16-distro-crimson-smithi/

**Comment by Matan-B** [comment]
jenkins test api

**Review by Matan-B — CHANGES_REQUESTED** [review]
```
DEBUG 2024-06-17 16:08:49,590 [shard 0:main] osd - do_recover_missing reqid client.15504.0:11 check for recovery, 100:57260762:test-rados-api-smithi033-96583-4::foo:head
DEBUG 2024-06-17 16:08:49,590 [shard 0:main] osd - do_recover_missing reqid client.15504.0:11 nothing to recover 100:57260762:test-rados-api-smithi033-96583-4::foo:head
ERROR 2024-06-17 16:08:49,590 [shard 0:main] none - /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4318-g0ad4b3f4/rpm/el9/BUILD/ceph-19.0.0-4318-g0ad4b3f4/src/crimson/common/operation.h:620 : In function 'void crimson::PipelineHandle::enter_sync(T&, typename T::BlockingEvent::Trigger<OpT>&&) [with OpT = crimson::osd::ClientRequest; T = crimson::osd::ClientRequest::PGPipeline::RecoverMissingLockOBC; typename T::BlockingEvent::Trigger<OpT> = crimson::BlockerT<crimson::osd::ClientRequest::PGPipeline::RecoverMissingLockOBC>::BlockingEvent::Trigger<crimson::osd::ClientRequest>; typename T::BlockingEvent = crimson::BlockerT<crimson::osd::ClientRequest::PGPipeline::RecoverMissingLockOBC>::BlockingEvent]', ceph_assert(%s)
!ret.has_value()
Aborting on shard 0.
```
from osd.0 and 3 in https://qa-proxy.ceph.com/teuthology/matan-2024-06-17_15:38:16-crimson-rados-wip-matanb-crimson-testing-june-16-distro-crimson-smithi/7759890/remote/smithi033/log/

Full run:
https://pulpito.ceph.com/matan-2024-06-17_15:38:16-crimson-rados-wip-matanb-crimson-testing-june-16-distro-crimson-smithi/

File: src/crimson/osd/osd_operations/client_request.cc [hunk 1]
```diff
@@ -286,8 +286,6 @@ ClientRequest::recover_missing_snaps(
   ObjectContextRef head,
   std::set<snapid_t> &snaps)
 {
-  co_await ihref.enter_stage<interruptor>(
-    client_pp(*pg).recover_missing_snaps, *this);
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/client_request.cc:290]: Is this change intentional or it this still WIP? If so, can you please explain? Same for `recover_missing_lock_obc`

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:290]: Yes, it's intentional. The reason is:

1. the pipeline phase in which the obc loading is done has to be entered "synchronouly", however, the leaving of an `OrderedConcurrentPhase` can't be so, which is the cause of the failure in https://github.com/ceph/ceph/pull/57966#pullrequestreview-2124738667
2. on the other hand, the whole missing object/snaps recovery procedure can and should be a `OrderedConcurrentPhase`, so I made the whole procedure encapsulated in the same concurrent `recover_missing` phase

Comment (Matan-B) [reply src/crimson/osd/osd_operations/client_request.cc:290]: This should probably be part of the commit message.
The stages removed here `recover_missing_lock_obc` and `recover_missing_snaps` has no other users and can be removed entirely, right?

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:290]: > This should probably be part of the commit message.
> The stages removed here `recover_missing_lock_obc` and `recover_missing_snaps` has no other users and can be removed entirely, right?

Yes, will do.

**Review by Matan-B — APPROVED** [review]
Commits can be probably squashed

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-07-07_12:22:32-crimson-rados-wip-matanb-crimson-testing-57966-distro-crimson-smithi/

**Comment by Matan-B** [comment]
```
2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stdout:185: write (excl) oid 49 current snap is 10
2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stdout:181:  finishing write tid 1 to smithi05334528-20
2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stderr:Error: finished tid 1 when last_acked_tid was 3
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4747-g5100804c/rpm/el9/BUILD/ceph-19.0.0-4747-g5100804c/src/test/osd/RadosModel.h: In function 'virtual void WriteOp::_finish(TestOp::CallbackInfo*)' thread 7f9e0e397640 time 2024-07-07T13:52:09.678182+0000
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4747-g5100804c/rpm/el9/BUILD/ceph-19.0.0-4747-g5100804c/src/test/osd/RadosModel.h: 1042: ceph_abort_msg("abort() called")
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: ceph version 19.0.0-4747-g5100804c (5100804cc9d9e0601eec0e579412ccd8666747f9) squid (dev)
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xcc) [0x7f9e17072c7e]
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 2: (WriteOp::_finish(TestOp::CallbackInfo*)+0x1bc) [0x44e0e2]
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 3: (TestOp::finish(TestOp::CallbackInfo*)+0xe) [0x45fda8]
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 4: (write_callback(void*, void*)+0x1c) [0x45fde3]
```
https://pulpito.ceph.com/matan-2024-07-07_12:22:32-crimson-rados-wip-matanb-crimson-testing-57966-distro-crimson-smithi/

**Review by Matan-B — CHANGES_REQUESTED** [review]
```
2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stdout:185: write (excl) oid 49 current snap is 10
2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stdout:181:  finishing write tid 1 to smithi05334528-20
2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stderr:Error: finished tid 1 when last_acked_tid was 3
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4747-g5100804c/rpm/el9/BUILD/ceph-19.0.0-4747-g5100804c/src/test/osd/RadosModel.h: In function 'virtual void WriteOp::_finish(TestOp::CallbackInfo*)' thread 7f9e0e397640 time 2024-07-07T13:52:09.678182+0000
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4747-g5100804c/rpm/el9/BUILD/ceph-19.0.0-4747-g5100804c/src/test/osd/RadosModel.h: 1042: ceph_abort_msg("abort() called")
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: ceph version 19.0.0-4747-g5100804c (5100804cc9d9e0601eec0e579412ccd8666747f9) squid (dev)
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xcc) [0x7f9e17072c7e]
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 2: (WriteOp::_finish(TestOp::CallbackInfo*)+0x1bc) [0x44e0e2]
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 3: (TestOp::finish(TestOp::CallbackInfo*)+0xe) [0x45fda8]
2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 4: (write_callback(void*, void*)+0x1c) [0x45fde3]
```
https://pulpito.ceph.com/matan-2024-07-07_12:22:32-crimson-rados-wip-matanb-crimson-testing-57966-distro-crimson-smithi/

**Comment by xxhdx1985126** [comment]
> ```
> 2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stdout:185: write (excl) oid 49 current snap is 10
> 2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stdout:181:  finishing write tid 1 to smithi05334528-20
> 2024-07-07T13:52:09.683 INFO:tasks.rados.rados.0.smithi053.stderr:Error: finished tid 1 when last_acked_tid was 3
> 2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4747-g5100804c/rpm/el9/BUILD/ceph-19.0.0-4747-g5100804c/src/test/osd/RadosModel.h: In function 'virtual void WriteOp::_finish(TestOp::CallbackInfo*)' thread 7f9e0e397640 time 2024-07-07T13:52:09.678182+0000
> 2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4747-g5100804c/rpm/el9/BUILD/ceph-19.0.0-4747-g5100804c/src/test/osd/RadosModel.h: 1042: ceph_abort_msg("abort() called")
> 2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: ceph version 19.0.0-4747-g5100804c (5100804cc9d9e0601eec0e579412ccd8666747f9) squid (dev)
> 2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xcc) [0x7f9e17072c7e]
> 2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 2: (WriteOp::_finish(TestOp::CallbackInfo*)+0x1bc) [0x44e0e2]
> 2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 3: (TestOp::finish(TestOp::CallbackInfo*)+0xe) [0x45fda8]
> 2024-07-07T13:52:09.684 INFO:tasks.rados.rados.0.smithi053.stderr: 4: (write_callback(void*, void*)+0x1c) [0x45fde3]
> ```
> 
> https://pulpito.ceph.com/matan-2024-07-07_12:22:32-crimson-rados-wip-matanb-crimson-testing-57966-distro-crimson-smithi/

@Matan-B I've just fixed the issue that caused the error, please take a look again, thanks:-)

File: src/crimson/osd/osd_operations/client_request.cc [hunk 2]
```diff
@@ -329,6 +329,8 @@ ClientRequest::process_op(
     }
   }
 
+  co_await ihref.enter_stage<interruptor>(
+    client_pp(*pg).check_already_complete, *this);
```
Comment (cyx1231st) [inline src/crimson/osd/osd_operations/client_request.cc:333]: Is it reasonable to merge `check_already_complete` with the follow-up `get_obc`? They are both ordered exclusive, following the same property. Seems to me introducing more mutexes in the critical IO path are unnecessary.

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:333]: I've no problem with merge those two phases, what do you think? @Matan-B

Comment (Matan-B) [reply src/crimson/osd/osd_operations/client_request.cc:333]: We can merge two sequential ordered exclusive stages. The only thing I'm afraid of is readability since we try to squeeze two logical stages into a single one.
Instead of reusing `get_obc` here, what do you think about introducing the merged stage as a new one? We can name it  `check_already_complete_and_get_obc`. Not the best name but future readers will be able to understand why we enter it before checking for already complete.
We should probably also add a comment explaining that the new stage combines two logical stages to avoid mutex usage overhead between sequential ordered exclusive stages.

Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc]: Sorry for the late response, the comment is addressed, I kept the original `get_obc` phase as other requests, like snaptrim, need it. Please take a look, again. Thanks:-)

File: src/crimson/osd/osd_operations/client_request.cc [hunk 3]
```diff
@@ -321,8 +317,6 @@ ClientRequest::process_op(
     std::set<snapid_t> snaps = snaps_need_to_recover();
     if (!snaps.empty()) {
       // call with_obc() in order, but wait concurrently for loading.
```
Comment (cyx1231st) [inline src/crimson/osd/osd_operations/client_request.cc:319]: This comment is stale.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 4]
```diff
@@ -355,7 +351,6 @@ ClientRequest::process_op(
 
   DEBUGDPP("{}.{}: not completed, entering get_obc stage",
 	   *pg, *this, this_instance_id);
-  co_await ihref.enter_stage<interruptor>(client_pp(*pg).get_obc, *this);
 
   DEBUGDPP("{}.{}: entered get_obc stage, about to wait_scrub",
 	   *pg, *this, this_instance_id);
```
Comment (cyx1231st) [inline src/crimson/osd/osd_operations/client_request.cc:356]: The two logs above need adjustment accordingly.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 5]
```diff
@@ -329,6 +329,8 @@ ClientRequest::process_op(
     }
   }
 
+  co_await ihref.enter_stage<interruptor>(
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/client_request.cc:332]:
```
/**
  * The previous stage of recover_missing is a concurrent phase.
  * Checking for already_complete requests must done exclusively.
  * Since get_obc is also an exclusive stage, we can merge both stages into
  * a single stage and avoid stage switching overhead.
  */
```

**Comment by xxhdx1985126** [comment]
Done:-)

File: src/crimson/osd/osd_operations/client_request.cc [hunk 6]
```diff
@@ -368,11 +375,7 @@ ClientRequest::process_op(
     co_return;
   }
 
-  DEBUGDPP("{}.{}: not completed, entering get_obc stage",
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/client_request.cc:377]: nit: let's add "  `DEBUGDPP("{}.{}: not completed, entering check_already_complete_get_obc stage" ... )` instead.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 7]
```diff
@@ -350,6 +349,14 @@ ClientRequest::process_op(
     }
   }
 
+  /**
+  * The previous stage of recover_missing is a concurrent phase.
+  * Checking for already_complete requests must done exclusively.
+  * Since get_obc is also an exclusive stage, we can merge both stages into
+  * a single stage and avoid stage switching overhead.
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/client_request.cc:356]: nit: indentation space missing

**Review by Matan-B — APPROVED** [review]

File: src/crimson/osd/osd_operations/client_request.cc (hunk 6 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:377]: Um, we are actually going to enter "wait_scrub" phase here. I added the corresponding logs below.

File: src/crimson/osd/osd_operations/client_request.cc (hunk 7 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc]: Done, sorry, I didn't notice this indentation issue.

File: src/crimson/osd/osd_operations/client_request.cc (hunk 6 above)
Comment (Matan-B) [reply src/crimson/osd/osd_operations/client_request.cc:377]: I meant adding log before entering `check_already_complete_get_obc`
DEBUGDPP("entering check_already_complete_get_obc stage" ... )
Shouldn't block just a suggestion :)

**Comment by Matan-B** [comment]
jenkins test make check

File: src/crimson/osd/osd_operations/client_request.cc (hunk 6 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:377]: Done:-)

**Comment by cyx1231st** [comment]
jenkins test make check

**Review by cyx1231st — APPROVED** [review]

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-aug-12

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-aug-14

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-08-14_16:27:16-crimson-rados-wip-matanb-crimson-testing-aug-14-distro-crimson-smithi/

7855673, 7855704 - infra
7855680 - new, untreated https://tracker.ceph.com/issues/67568
7855707 - https://tracker.ceph.com/issues/66767
